### DISCUSSION TITLE
How to make a Binding to comparison result of two object?

### DISCUSSION THREAD

**Discussion by WarWithinMe** [AvaloniaUI/Avalonia#13327] [Q&A]
In my ViewModel, I have a list of items `Items`, it also has a property `SelectedItem` indicating which item is selected.
In my view, I display the items using ItemsControl, and I would like to add a class `Selected` when the item ReferenceEquals to the `SelectedItem`.
```
<ItemsControl ItemsSource="{Binding Items}">
    <ItemsControl.ItemsPanel>
         <ItemsPanelTemplate><StackPanel /></ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>
    <ItemsControl.ItemTemplate>
        <DataTemplate>
            <Button Classes="MyButton" Classes.Selected="{Binding $parent[ItemsControl].((vm:ViewModel)DataContext).SelectedItem, Converter={x:Static vm:IsSelectedConverter.shared}, ConverterParameter={Binding}}">
            </Button>
         </DataTemplate>
     </ItemsControl.ItemTemplate>
</ItemsControl>
```

So I tried using a converter which returns true if the SelectedItem equals to the item. However, in the IValueConverter.Convert(), the ConverterParameter is a CompiledBindingExtension object, and I don't know how to get the real value of item.

And it's not easy for me to just add an `IsSelected` property to the item, because I need to display the list in several windows, and each window might have its own selection.

Any advice will be appreciated. Thanks.

**Comment by jp2masa** [comment]
`ConverterParameter` can't be a binding. You can use `MultiBinding` with `IMultiValueConverter` instead.

**Reply by WarWithinMe** [reply]
OK, but how to bind to a ClassName with MultiBinding ?

This is not gonna work:
```
<Button.Classes.Selected>
    <MultiBinding Converter="{StaticResource WidthToRenderTransformConverter}">
        <Binding Path="$parent[ItemsControl].((vm:ViewModel)DataContext).SelectedItem"/>
        <Binding Path="."/>
    </MultiBinding>
</Button.Classes.Selected>
```

**Reply by jp2masa** [reply]
You're right, I missed that detail. Did you try creating a `MultiBinding` in resources and then using `Classes.Selected="{StaticResource MyMultiBinding}"`?

**Reply by WarWithinMe** [reply]
It doesn't work. Here's what I did:
```
<ItemsControl>
<ItemsControl.Resources>
    <SolidColorBrush x:Key="MySolidBrush" Color="Red" />
    <Binding x:Key="MyBinding" />
    <MultiBinding x:Key="MyMultiBinding" Converter="{x:static WidthToRenderTransformConverter}">
        <Binding Path="$parent[ItemsControl].((vm:ViewModel)DataContext).SelectedItem"/>
        <Binding Path="."/>
    </MultiBinding>
</ItemsControl.Resources>
<ItemsControl.ItemTemplate>
        <DataTemplate>
            <Button Classes="MyButton" Classes.Selected="{StaticResource MyMultiBinding}" Background="{Binding MySolidBrush}">
            </Button>
         </DataTemplate>
</ItemsControl.ItemTemplate>
</ItemsControl>
```

#### Test with MyMultiBinding 
1. With `Classes.Selected="{StaticResource MyMultiBinding}"`, the compiler fails to compile throwing the error "Unable to convert {x:Null} to System.Runtime:System.Boolean".
2. With `Classes.Selected="{DynamicResource MyMultiBinding}"`, the compiler compiles, but it does nothing. I added a breakpoint in MultiBinding's constructor, it seems to me that MultiBinding is being constructed, but its associated converter never gets called.

#### Test with MyBinding
It behaves just like MultiBinding, (StaticResource doesn't compile, DynamicResource compiles but does nothing), except that the source editor now shows a warning stating "Resource 'MyBinding' is not found".

#### Test with MySolidBrush
The button's background changes to Red. So it indicates I'm correctly using Resources in this sample.

**Comment by WarWithinMe** [comment]
It also seems like I can't create a binding to the current DataContext itself in MultiBinding. I can do something like <Button Classes="MyButton" Classes.Selected="{Binding}"/>. The inlined `{Binding}` resolves to the DataContext.

In the <MultiBinding> below, I can read the value of the first binding in the Converter, but not the second binding.

```
<Button.Classes>
    <MultiBinding Converter="{StaticResource WidthToRenderTransformConverter}">
        <Binding Path="$parent[ItemsControl].((vm:ViewModel)DataContext).SelectedItem"/>
        <Binding Path="."/>  // This doesn't work. Nor `<Binding />` does.
    </MultiBinding>
</Button.Classes>
```

**Comment by WarWithinMe** [comment]
For now, I'm using Behavoir to solve this problem:

```
public class ToggleClassBehavior : Behavior<Control>
{
    public static readonly DirectProperty<ToggleClassBehavior, object?> Param1Property =
        AvaloniaProperty.RegisterDirect<ToggleClassBehavior, object?>(nameof(Param1), b => b.Param1, (b, v) => b.Param1 = v);
    
    public static readonly DirectProperty<ToggleClassBehavior, object?> Param2Property =
        AvaloniaProperty.RegisterDirect<ToggleClassBehavior, object?>(nameof(Param2), b => b.Param2, (b, v) => b.Param2 = v);
    
    public static readonly DirectProperty<ToggleClassBehavior, IMultiValueConverter?> ConverterProperty =
        AvaloniaProperty.RegisterDirect<ToggleClassBehavior, IMultiValueConverter?>(nameof(Converter), b => b.Converter, (b, v) => b.Converter = v);
    
    /// <summary>
    /// A converter that should return true if a class should be added.
    /// </summary>
    public IMultiValueConverter? Converter
    {
        get => _converter;
        set
        {
            _converter = value;
            ToggleClass();
        }
    }

    public object? Param1
    {
        get => _param1;
        set
        {
            _param1 = value;
            ToggleClass();
        }
    }

    public object? Param2
    {
        get => _param2;
        set
        {
            _param2 = value;
            ToggleClass();
        }
    }

    public string Class
    {
        get => _class;
        set
        {
            _class = value;
            ToggleClass();
        }
    }

    private IMultiValueConverter? _converter;
    private object?               _param1;
    private object?               _param2;
    private string                _class = "";

    protected override void OnAttachedToVisualTree()
    {
        base.OnAttachedToVisualTree();
        ToggleClass();
    }

    private void ToggleClass()
    {
        if (_converter.Convert(new List<object?> { _param1, _param2 }, typeof(bool), null, CultureInfo.CurrentCulture) is true)
        {
            AssociatedObject.Classes.Add(_class);
        }
        else
        {
            AssociatedObject.Classes.Remove(_class);
        }
    }
}
```

Usage:
```
<ItemsControl>
    <ItemsControl.ItemTemplate>
        <DataTemplate>
            <Button>
                <Interaction.Behaviors>
                    <myNameSpace:ToggleClassBehavior Class="Selected"
                        Param1="{Binding $parent[ItemsControl].((vm:ViewModel)DataContext).SelectedItem}"
                        Param2="{Binding}"
                        Converter="{x:Static myNameSpace:ReferenceEqualConverter}"/>
                </Interaction.Behaviors>
            </Button>
        </DataTemplate>
    </ItemsControl.ItemTemplate>
</ItemsControl>
```
It's not very flexible, but it does the trick.

**Comment by nsrahmad** [comment]
@WarWithinMe 
EDIT: I missed the part where you explained why `IsSelected` is not suitable, so never mind. but I will leave the answer as is.

This looks way too complicated for me. 
When I was in similar situation, I gave each item an `IsSelected` property. I was using CommunityToolkit.Mvvm package, which also generates partial methods like `OnSelectedCircleChanging` and `OnSelectedCircleChanged` for a Property named `SelectedCircle`. So I added something like in the VM :
```csharp
 partial void OnSelectedCircleChanging(CircleViewModel? value)
  {
        // The property is about to change, but still referring to old property value
        if (SelectedCircle == null) return;
        SelectedCircle.IsSelected = false;
  }
 
partial void OnSelectedCircleChanged(CircleViewModel? value) => value!.IsSelected = true;
```
This effectively acheives what you want to do, and is much simpler.
